db_util.go

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2022 Red Hat, Inc.
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ */
+
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	c "nav/constants"
+)
+
+type Datasource interface {
+	init(arg interface{}) (err error)
+	GetExploredSubsystemByName(subs string) string
+	getSuccessorsById(symbolId int, instance int) ([]entry, error)
+	getSubsysFromSymbolName(symbol string, instance int) (string, error)
+	sym2num(symb string, instance int) (int, error)
+	symbSubsys(symblist []int, instance int) (string, error)
+	getEntryById(symbolId int, instance int) (entry, error)
+}
+
+const SUBSYS_UNDEF = "The REST"
+
+// Parent node.
+type node struct {
+	subsys     string
+	symbol     string
+	sourceRef  string
+	addressRef string
+}
+type adjM struct {
+	l node
+	r node
+}
+
+type entry struct {
+	symbol     string
+	fn         string
+	sourceRef  string
+	addressRef string
+	subsys     []string
+	symId      int
+}
+
+type edge struct {
+	sourceRef  string
+	addressRef string
+	caller     int
+	callee     int
+}
+
+// Return id an item is already in the list.
+func notIn(list []int, v int) bool {
+
+	for _, a := range list {
+		if a == v {
+			return false
+		}
+	}
+	return true
+}
+
+// Removes duplicates resulting by the exploration of a call tree.
+func removeDuplicate(list []entry) []entry {
+
+	sort.SliceStable(list, func(i, j int) bool { return list[i].symId < list[j].symId })
+	allKeys := make(map[int]bool)
+	var res []entry
+	for _, item := range list {
+		if _, value := allKeys[item.symId]; !value {
+			allKeys[item.symId] = true
+			res = append(res, item)
+		}
+	}
+	return res
+}
+
+// Checks if a given function needs to be explored.
+func notExcluded(symbol string, excluded []string) bool {
```

Comment:
The function `regexp.MatchString("foo", "foobar")` returns a match because that is how regular expressions work. When I specify a field as a regular expression, I assume that the user is familiar with how regular expressions work. I do not want to limit the potential of regular expressions simply because a user may misuse them. The name `notExcluded` was chosen because it sounded appropriate to me, given that the `excluded.*` enums are utilized.